rohmu/util.py

```diff
@@ -71,3 +74,131 @@ def get_total_size_from_content_range(content_range: str) -> Optional[int]:
     length = content_range.rsplit("/", 1)[1]
     # RFC 9110 section 14.4 specifies that the * can be returned when the total length is unknown
     return int(length) if length != "*" else None
+
+
+class BinaryStreamsConcatenation:
+    """Concatenate a sequence of binary streams.
+    The concatenation only allows for the read() call.
+    """
+
+    def __init__(self, files: Iterable[BinaryIO]) -> None:
+        self._iter_files = iter(files)
+        self._current_file: Optional[BinaryIO] = None
+
+    def _read_chunk(self, size: int = -1) -> bytes:
+        if self._current_file is None:
+            self._current_file = next(self._iter_files, None)
+            if self._current_file is None:
+                return b""
+        data = self._current_file.read(size) if size > 0 else self._current_file.read()
+        if not data:
+            self._current_file.close()
+            self._current_file = next(self._iter_files, None)
+        return data
+
+    def read(self, size: int = -1) -> bytes:
+        result = BytesIO()
+        size_left = size
+        while True:
+            chunk = self._read_chunk(size_left)
+            if not chunk and self._current_file is None:
+                # we finished reading all files
+                break
+            result.write(chunk)
+            size_left -= len(chunk)
+            if size > 0 and size_left == 0:
+                # we finished reading the amount requested
+                break
+
+        return result.getvalue()
+
+
+class ProgressStream(BinaryIO):
```

Comment:
BinaryIO is ugly... here we just want something that wraps a file-like object and can report how much bytes it has read.